torch/csrc/autograd/engine.cpp

```diff
@@ -1154,40 +1175,46 @@ void GraphTask::init_to_execute(Node& graph_root, const edge_list& outputs, bool
       return nullptr;
     }
   };
+
+  auto nodeShouldExecute = [this](Node *fn) {
+    auto it = exec_info_.find(fn);
+    return it != exec_info_.end() && it->second.should_execute();
+  };
+
   std::vector<Frame> stack;
   std::unordered_set<Node*> seen;
-  for (const auto & input : graph_root.next_edges()) {
-    if (!input.function.get()) continue;
-    if (seen.count(input.function.get()) > 0) continue;
-    stack.emplace_back(input.function.get());
-    while (!stack.empty()) {
-      auto &frame = stack.back();
-      if (Node *next_fn = frame.get_next_fn()) {
-        if (/* bool unseen = */ seen.emplace(next_fn).second) {
-          stack.emplace_back(next_fn);
-          continue; // recurse
-        }
-      } else {
-        // NB: if we were using real recursion we could have saved some lookups
-        // using a return value from recursive call. It would make this manually unrolled
-        // version a lot more complicated, so I skipped that.
-        const auto & next_edges = frame.fn_->next_edges();
-        const bool needed = std::any_of(
-            next_edges.begin(), next_edges.end(), [&](const Edge& edge) {
-              auto it = exec_info_.find(edge.function.get());
-              return it != exec_info_.end() && it->second.should_execute();
-            });
-        exec_info_[frame.fn_].needed_ |= needed;
-        stack.pop_back();
+  stack.emplace_back(&graph_root);
+  exec_info_.emplace(stack.back().fn_, ExecInfo());
+
+  while (!stack.empty()) {
+    auto &frame = stack.back();
+    const auto fn = frame.fn_;
+
+    Node *child_fn = nullptr;
+    while((child_fn = frame.get_next_fn()) && !seen.emplace(child_fn).second) {
+      // (1) next child exists AND has already been seen
+      if (nodeShouldExecute(child_fn)) {
+        exec_info_[fn].needed_ = true;
       }
     }
-  }
 
-  exec_info_[&graph_root].needed_ |= std::any_of(
-      graph_root.next_edges().begin(), graph_root.next_edges().end(), [&](const Edge& edge) {
-        auto it = exec_info_.find(edge.function.get());
-        return it != exec_info_.end() && it->second.should_execute();
-      });
+    if (child_fn) {
+      // (2) next child exists but has not been seen
+      const size_t max_nb_collect_next_edges = 10; // ??
+      if (child_fn->actual_sequence_nr() >= min_seq_nr - max_nb_collect_next_edges) {
```

Comment:
Don't recurse if the child node is created earlier than the earliest created output. This is tricky though because AccumulateGrad is lazily materialized so that its sequence_nr is actually greater than the node that has it as its next edge. We can deal with this by subtracting by a small constant `max_nb_collect_next_edges`, but how do we determine this constant?